lib/puppet/provider/ec2_instance/v2.rb

```diff
@@ -46,10 +46,11 @@ def self.instance_to_hash(region, instance)
     end
     subnet_name = nil
     if instance.subnet_id
-      subnet_response = ec2_client(region).describe_subnets(subnet_ids: [instance.subnet_id])
-      subnet_name_tag = subnet_response.data.subnets.first.tags.detect { |tag| tag.key == 'Name' }
+      subnet_response = ec2_client(region).describe_subnets(filters: [
+        {name: 'subnet-id', values: [instance.subnet_id]}
+      ])
+      subnet_name = subnet_response.data.subnets.empty? ? nil : name_from_tag(subnet_response.data.subnets.first)
```

Comment:
I think the crux of my issue here is not correctly representing the underlying state of the resource, and then later making decisions based on this misrepresentation.  When code branches occur that are driven based on the incorrect state, this can lead to unpredictable behavior.

While it might seem to make sense that read-only properties on the resource are OK to omit based on transient failures, I think it actually makes the code much more difficult to reason about and later maintain. 

So I'm in favor of raising errors / failing fast when transient issues occur.  I would rather see the provider code fail as a result of AWS or other networking issues, rather than continue in a degraded state.